src/runtime/qurt_hvx.cpp

```diff
@@ -52,4 +53,214 @@ WEAK void halide_qurt_hvx_unlock_as_destructor(void *user_context, void * /*obj*
     halide_qurt_hvx_unlock(user_context);
 }
 
+/////////////////////////////////////////////////////////////////////////////
+//
+// halide_hexagon_prefetch_buffer_t
+//
+// Prefetch a multi-dimensional box (subset) from a larger array
+//
+//   dim        number of dimensions in array
+//   elem_size  size in bytes of one element
+//   num_elem   total number of elements in array
+//   buf        buffer_t describing box to prefetch
+//              (note: extent/stride/min are all assumed to be positive)
+//
+// Notes:
+//  - Prefetches can be queued up to 3 deep (MAX_PREFETCH)
+//  - If 3 are already pending, the oldest request is dropped
+//  - USR:PFA status bit is set to indicate that prefetches are in progress
+//  - A l2fetch with any subfield set to zero cancels all pending prefetches
+//  - The l2fetch starting address must be in mapped memory but the range
+//    prefetched can go into unmapped memory without raising an exception
+//
+// TODO: Opt: Generate more control code for prefetch_buffer_t directly in
+// TODO       Prefetch.cpp to avoid passing box info through a buffer_t
+// TODO       (which results in ~30 additional stores/loads)
+//
+#define MAX_PREFETCH    3
+#define MASK16          0xFFFF
+#define MIN(x,y)        (((x)<(y)) ? (x) : (y))
+#define MAX(x,y)        (((x)>(y)) ? (x) : (y))
+//
+// Prefetch debugging controls
+//#define DEBUG_PREFETCH       1  // Uncomment to enable prefetch debug
+
+#ifdef DEBUG_PREFETCH // Prefetch debug enabled
+
+// Only produce debug info in this range
+#define DBG_P_START          1  // First debug instance to display
+#define DBG_P_STOP           9  // Last debug instance to display
+//#define DBG_P_STOP           0xFFFFFFFF  // Last debug instance to display
+
+#define DBG_PREFETCH_V(...)  log_printf(__VA_ARGS__) // Verbose debug enabled
+//#define DBG_PREFETCH_V(...)  {} // Verbose debug disabled
+#define DBG_PREFETCH(...)    if ((dbg_cnt >= DBG_P_START) && \
+                                 (dbg_cnt <= DBG_P_STOP)) { \
+                                 log_printf(__VA_ARGS__); \
+                             }  // Normal range-based debug enabled
+
+#else // Prefetch debug disabled
+
+#define DBG_PREFETCH_V(...)  {} // Verbose debug disabled
+#define DBG_PREFETCH(...)    {} // Normal debug disabled
+
+#endif
+//
+WEAK int halide_hexagon_prefetch_buffer_t(const uint32_t dim, const int32_t elem_size,
+                const uint32_t num_elem, const buffer_t *buf)
+{
+    // Extract needed fields from buffer_t
+    unsigned char *addr   = buf->host;
+    const int32_t *extent = buf->extent;
+    const int32_t *stride = buf->stride;
+    const int32_t *min    = buf->min;
+    const uint32_t boxdim = dim;  // dimensions of entire box to prefetch
```

Comment:
I think boxdim should be removed, and the 'dim' parameter should be renamed to something more descriptive, e.g. dim_count.